spring-xd-dirt/src/main/java/org/springframework/xd/dirt/stream/StreamDeployer.java

```diff
@@ -80,4 +114,70 @@ protected String getDeploymentPath(StreamDefinition definition) {
 		return Paths.build(Paths.STREAM_DEPLOYMENTS, definition.getName());
 	}
 
+	/**
+	 * The migration code to run against existing stream definitions that don't have module definitions set already.
+	 * The following method will update the module definitions for those stream definitions.
+	 * See https://jira.spring.io/browse/XD-2854
+	 */
+	@PostConstruct
+	private void updateModuleDefinitions() {
+		if (this.zkConnection.getClient() != null) {
+			try {
+				CuratorFramework client = this.zkConnection.getClient();
+				if (client.checkExists().forPath(Paths.STREAMS) != null) {
+					Iterable<StreamDefinition> streamDefinitions = findAll();
+					for (StreamDefinition definition : streamDefinitions) {
+						setModuleDefinitions(client, definition);
+					}
+				}
+			}
+			catch (Exception e) {
+				logger.error("Exception migrating stream definitions. This migration is done when the existing " +
+						"stream definitions that don't have module definitions set. " + e);
+			}
+		}
+	}
+
+	/**
+	 * Set the module definitions for the given stream definition.
+	 * Module definitions are obtained by parsing the stream definition.
+	 *
+	 * @param client the curator framework client
+	 * @param definition the stream definition
+	 */
+	private void setModuleDefinitions(CuratorFramework client, StreamDefinition definition) {
+		String streamName = definition.getName();
+		String path = Paths.build(Paths.STREAMS, streamName);
+		try {
+			byte[] bytes = client.getData().forPath(path);
+			if (bytes != null) {
```

Comment:
If the stream node exists then it would never be empty as it will always at least have the stream definition with it.